src/components/layer-controller/LayerControllerSubscriber.js

```diff
@@ -6,53 +6,133 @@ import {
 import LayerController from './LayerController';
 import ImageAddButton from './ImageAddButton';
 import {
-  RASTER_ADD, LAYER_REMOVE, CLEAR_PLEASE_WAIT, METADATA_REMOVE,
+  RASTER_ADD, LAYER_REMOVE, CLEAR_PLEASE_WAIT, METADATA_REMOVE, LAYER_ADD, METADATA_ADD,
 } from '../../events';
 import { darkTheme } from './styles';
-
+import { DEFAULT_LAYER_PROPS } from './constants';
 
 const generateClassName = createGenerateClassName({
   disableGlobal: true,
 });
 
+async function initLoader(imageData) {
+  const {
+    type, url, metadata, requestInit,
+  } = imageData;
+  switch (type) {
+    case ('zarr'): {
+      const { dimensions, isPyramid, transform } = metadata;
+      const { scale = 0, translate = { x: 0, y: 0 } } = transform;
+      const loader = await createZarrLoader({
+        url, dimensions, isPyramid, scale, translate,
+      });
+      return loader;
+    }
+    case ('ome-tiff'): {
+      // Fetch offsets for ome-tiff if needed.
+      if ('omeTiffOffsetsUrl' in metadata) {
+        const { omeTiffOffsetsUrl } = metadata;
+        const res = await fetch(omeTiffOffsetsUrl, requestInit);
+        if (res.ok) {
+          const offsets = await res.json();
+          const loader = await createOMETiffLoader({
+            url,
+            offsets,
+            headers: requestInit,
+          });
+          return loader;
+        }
+        throw new Error('Offsets not found but provided.');
+      }
+      const loader = createOMETiffLoader({
+        url,
+        headers: requestInit,
+      });
+      return loader;
+    }
+    default: {
+      throw Error(`Image type (${type}) is not supported`);
+    }
+  }
+}
+
+function publishLayer({ loader, imageData, layerId }) {
+  PubSub.publish(LAYER_ADD, {
+    layerId,
+    loader,
+    layerProps: DEFAULT_LAYER_PROPS,
+  });
+  if (loader.getMetadata) {
+    PubSub.publish(METADATA_ADD, {
+      layerId,
+      layerName: imageData.name,
+      layerMetadata: loader.getMetadata(),
+    });
+  }
+}
+
 function LayerControllerSubscriber({ onReady, removeGridComponent }) {
   const [imageOptions, setImageOptions] = useState(null);
-  const [layers, setLayers] = useState([]);
+  const [layersAndLoaders, setLayersAndLoaders] = useState([]);
   const memoizedOnReady = useCallback(onReady, []);
 
   useEffect(() => {
-    function handleRasterAdd(msg, raster) {
-      setImageOptions(raster.images);
+    async function handleRasterAdd(msg, raster) {
+      // render_layers provides the order for rendering initially.
+      const { images, renderLayers } = raster;
+      setImageOptions(images);
+      if (!renderLayers) {
+        const layerId = String(Math.random());
+        // Midpoint of images list as default image to show.
+        const imageData = images[Math.floor(images.length / 2)];
+        const loader = await initLoader(imageData);
+        publishLayer({ loader, imageData, layerId });
+        setLayersAndLoaders([...layersAndLoaders, { layerId, imageData, loader }]);
```

Comment:
should use the callback option in `setLayersAndLoaders` rather than spreading the current state. This means the state won't be in the dependency array --> its safer and avoids race conditions.

```javascript
setLayersAndLoaders(prev => [...prev, { layerId, imageData, loader }]);
```